lisa/_assets/kmodules/sched_tp/perf_counters.c

```diff
@@ -0,0 +1,694 @@
+
+static const struct attribute *perfctr_attrs[] = {
+	&rawctr_attrib.attr,
+	&sampling_attrib.attr,
+	&max_events_attrib.attr,
+	NULL
+};
+
+static int perfctr_sysfs_init(struct perfctr_core *perf_data)
+{
+	int retval;
+
+	init_completion(&perf_data->kobj_unregister);
+	retval = kobject_init_and_add(&perf_data->kobj, &perfctr_kobj_type,
+				      kernel_kobj, "perf_counters");
+	if (!retval)
+		retval = sysfs_create_files(&perf_data->kobj, perfctr_attrs);
+	if(retval) {
+		wait_for_completion(&perf_data->kobj_unregister);
+		kobject_put(&perf_data->kobj);
+	}
+	return retval;
+}
+
+static void perfctr_sysfs_release(struct perfctr_core *perf_data)
+{
+	sysfs_remove_files(&perf_data->kobj, perfctr_attrs);
+	kobject_del(&perf_data->kobj);
+	kobject_put(&perf_data->kobj);
+	wait_for_completion(&perf_data->kobj_unregister);
+	return;
+}
+#else
+static int perfctr_sysfs_init(struct perfctr_core *perf_data) { return 0; }
+static void perfctr_sysfs_release(struct perfctr_core *perf_data) { }
+#endif
+
+static void perfctr_pmu_discover(struct perfctr_core *perf_data)
+{
+	struct perf_event *event;
+	cpumask_var_t active_mask;
+	int cpu;
+
+	/*
+	 * This is absolutely loathsome but there seems to be no other way
+	 * to poke relevant pmu driver for details so, there it is ....
+	 */
+	struct perf_event_attr attr = {
+		.type		= PERF_TYPE_HARDWARE,
+		.size		= sizeof(struct perf_event_attr),
+		.pinned		= 1,
+		.disabled	= 1,
+		.config		= PERF_COUNT_HW_CPU_CYCLES,
+	};
+
+	perf_data->max_nr_events = MAX_PERF_COUNTERS;
+
+	if (!IS_ENABLED(CONFIG_ARM_PMU))
+		return;
+
+	if (!zalloc_cpumask_var(&active_mask, GFP_KERNEL))
+		return;
+
+	for_each_possible_cpu(cpu) {
+
+		if (cpumask_test_cpu(cpu, active_mask))
+		    continue;
+
+		event = perf_event_create_kernel_counter(&attr, cpu, NULL ,
+							 NULL, NULL);
+
+		if (IS_ERR(event)) {
+			pr_err("Failed to create an event (cpu%d) while discovery\n",
+				cpu);
+			break;
+		}
+
+		if (event->pmu) {
+			struct arm_pmu *pmu = to_arm_pmu(event->pmu);
+
+			perf_data->max_nr_events = min_t(unsigned int,
+							 perf_data->max_nr_events,
+							 pmu->num_events);
+
+			cpumask_or(active_mask, active_mask, &pmu->supported_cpus);
+
+		}
+		perf_event_release_kernel(event);
+
+		if (cpumask_equal(active_mask, cpu_possible_mask))
+				break;
+	}
+	free_cpumask_var(active_mask);
+	return;
+}
+
+static int perfctr_disable(struct feature *feature);
+
+static int perfctr_enable(struct feature *feature)
+{
+	struct perfctr_core *perf_data;
+	struct perfctr_match match;
+	int cpu;
+
+	if (!IS_ENABLED(CONFIG_HW_PERF_EVENTS)) {
+		pr_err("Missing support for HW performance event counters\n");
+		return 1;
+	}
+
+	if (ENABLE_FEATURE(__worqueue))
+		return 1;
+
+	perf_data = kzalloc(sizeof(*perf_data), GFP_KERNEL);
+	if (!perf_data) {
+		DISABLE_FEATURE(__worqueue);
+		return 1;
+	}
+
+	if (init_srcu_struct(&perf_data->srcu)) {
+		kfree(perf_data);
+		DISABLE_FEATURE(__worqueue);
```

Comment:
In the context of the previous comment, this will break refcounting: `perfctr_disable()` will always disable run `DISABLE_FEATURE(__worqueue)` provided `perf_data != NULL`, but that block also runs `DISABLE_FEATURE(__worqueue);` with `perf_data != NULL`.